src/Features/Core/Portable/UseCompoundAssignment/AbstractUseCompoundAssignmentDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,202 @@
+            => context.RegisterSyntaxNodeAction(AnalyzeAssignment, GetAnalysisKind());
+
+        private void AnalyzeAssignment(SyntaxNodeAnalysisContext context)
+        {
+            var cancellationToken = context.CancellationToken;
+            var assignment = (TAssignmentSyntax)context.Node;
+
+            var syntaxTree = assignment.SyntaxTree;
+            var optionSet = context.Options.GetDocumentOptionSetAsync(syntaxTree, cancellationToken).GetAwaiter().GetResult();
+            if (optionSet == null)
+            {
+                return;
+            }
+
+            var option = optionSet.GetOption(CodeStyleOptions.PreferCompoundAssignment, assignment.Language);
+            if (!option.Value)
+            {
+                // Bail immediately if the user has disabled this feature.
+                return;
+            }
+
+            _syntaxFacts.GetPartsOfAssignmentExpressionOrStatement(assignment, 
+                out var assignmentLeft, out var assignmentToken, out var assignmentRight);
+
+            // has to be of the form:  a = b op c
+            // op has to be a form we could convert into op=
+            if (!(assignmentRight is TBinaryExpressionSyntax binaryExpression))
+            {
+                return;
+            }
+
+            var binaryKind = GetKind(binaryExpression.RawKind);
+            if (!_binaryToAssignmentMap.ContainsKey(binaryKind))
+            {
+                return;
+            }
+
+            _syntaxFacts.GetPartsOfBinaryExpression(binaryExpression,
+                out var binaryLeft, out _);
+            
+            // has to be of the form:   expr = expr op ...
+            if (!_syntaxFacts.AreEquivalent(assignmentLeft, binaryLeft))
+            {
+                return;
+            }
+
+            // Syntactically looks promising.  But we can only safely do this if 'expr'
+            // is side-effect-free since we will be changing the number of times it is
+            // execute from twice to once.
+            var semanticModel = context.SemanticModel;
+            if (!IsSideEffectFree(assignmentLeft, semanticModel, cancellationToken))
+            {
+                return;
+            }
+
+            context.ReportDiagnostic(DiagnosticHelper.Create(
+                Descriptor,
+                assignmentToken.GetLocation(),
+                option.Notification.Severity,
+                additionalLocations: ImmutableArray.Create(assignment.GetLocation()),
+                properties: null));
+        }
+
+        private bool IsSideEffectFree(
+            SyntaxNode expr, SemanticModel semanticModel, CancellationToken cancellationToken)
+        {
+            if (expr == null)
+            {
+                return false;
+            }
+
+            // it basically has to be of the form "a.b.c", where all components are locals,
+            // parameters or fields.  Basically, nothing that can cause arbitrary user code
```

Comment:
seems good one to put in ISemanticFactService?

Reply:
it woudl be a little bit tough since it's so specific to this feature.  for example.  x.Prop is not side-effect free if Prop has side effects. However, x.Prop is fine to fixup here if there's just a single read/write.  

Do you have examples of other features that would benefit from this?

Reply:
what if .Prop is a ref returning property? then it seems we can't convert it because two calls to the getter would become one call

Reply:
@Neme12 Good catch!  will address that.

Reply:
it feels like genuinely good helper method. for example, we have extract to local variable or something which find all matching expression and change to local variable. that seems can have same issue as this. 

by the way, thinking more, it might be better to just offer fix even with side effect and show in preview that it might have side effect rather than not offer at all due to a chance which might not actually there in practice most of time.

Reply:
That can be done when the 'refactoring' side of this feature goes in.  HOwever, i don't want to do that work until https://github.com/dotnet/roslyn/pull/28768 goes in.  Right now it's just too onerous to write an analyzer/fixer/refactoring.  I want that PR to be accepted so that it's then trivial to create the single entity that does all this properly.

Thanks!:)